File: Python/Python Learning FIles/basics.py
# PyCharm


# Basics of Python

# 1. Data#		##########################################################################################################################################################
def basicData():
    # 1.1 variables
    string1 = "hi"  # makes a string with 'hi'
    int = 1  # makes a integer 1
    float = 3.14  # makes a float 3.14

    # a float is a decimal, a integer is a whole number
    answer = True  # a boolean, is a true or false variable
    replay = False
    nothing = None  # 'None' is an absance(blank) of data, because when making a variable you need somehting

    var1, var2, var3 = 1, 2, 3  # making multiple variables on one line, var1 goes to 1, var2 goes to 2 etc..
    print(int)


##### Naming ##############################################
# this is how I (Drake Thomas) would name things, everybody does it differently
"""
-Overview-
x, variable, thing = 12/'hi'/1.2
list2, dict1, namesOfPeople, foodToGet = []/{}/()
class Class, ClassOfStuff, ASimpleClass:
def func1/function_Does_Something/a_Quick_Func():

-Variables/lists/dict/etc-
integer = 12  --  one word
x = 12  --  should only use in enclosed loops, for global variables that you'll use a lot use comprehensive names
xy = 12
thingy = 'hi'  --  a quick variable you might not use a lot
sayHi = 'hello' --  something you might use a l ot
aVariableOfStuff = 'this is a variable' --  a very import variable that you'll call a lot and need to know what it is
list1 = []
aList = []
listWithData = []
aListOfNames = []
dict1 = {}
dictOfHomesAddrs = {}  -- a longer name. a dictionary of home addresses

-Functions-
def func1():  --  a quick function
def func_1():
def a_Function():
def function_That_Does_Stuff():  --   a important function that you might call a lot, or need to know what it does
def get_In_foFrom_Dict1():  --  function name that tells you what it does
def x():  --  should not use this. since it's usually a variable etc
function_var  --  all lowercase with _
-Class-
class Class1:  --  should be capitlized.
class Thingy_ThatDoes_Things:
class A_Quick_Class:
ClassVar  --  class variable, all uppercase, a class method var would be the same as a regular function name
Class_Function_Stuff()  --  a class function/method

-Other-
_obj  --  semi-private object
_function  --  this is usually used in a module file to hide the function, which means it's not meant to be used by the user, only the file. You can of course call the function but there's no point, only the file uses it
_var  --  same as the _function, this is suppose to be a hidden variable
__obj  --  super-private interpreter will usually do something with this
__name__  --  reserved for other python keywords/builtin/methods/attr/etc

"""


def printData():
    # 1.2 printing
    print("hi")  # prints text to console
    print("bye")  # prints 'bye' on a new-line

    print("hi " + "\nbye")  # prints 'hi' 'bye' on separated lines, like the two prints above just on one line

    print("yo", 1)  # prints 'yo' and then 1, the ',' is to print other stuff
    print("yo " + "hey")  # concatenates the two words, you need to add spaces in the string so it doesn't run on

    print('lol' * 10)  # prints out string 10 times

    print('1'), print('2')
    # using two print statements on the same line, but it doesn't actully print 1 and two on same line
    # you can usually do multiple things on the same line, sometimes not. Experiment


def lTSD():
    # list, tuple, set, dictionary

    # 1.3 lists, tuples, sets, and dictionaries
    list1 = [1, 2, 3, 'hi', 'bye']  # makes a list 'list1' (editable)
    list2 = [1, 2, 3, [4, 5, [6, 7], 8], 9]  # nested lists
    # a list in a list well be treated as one item when calling.
    # e.g. print(list[3]) this well print out all of the in the nested list that was made in position 4 >
    # [4,5,[6,7],8]

    list1.append('goodbye')  # adds 'goodbye' to the end of list1
    list1.pop(1)  # removes 1 from list1

    print('list1:', list1[1:3])  # prints items 2 - 4, :2, 5:, :, -3:, :-3
    print('list2', list2[4][1])
    # this well print out 5, since you call the list at index 4 and then the item in the list at index 1

    # Tuples
    tuple1 = (4, 5, 6, "foo", "bar")  # makes a tuple 'tuple1' (non-editiable)
    # you can't edit tuples

    # Sets
    set1 = {1, 1, 2, 2, 3, 3, 3, 'four', 'four', 'five', 'five'}
    # same as a list but it takes out duplicates
    print('set1:', set1)  # prints out set

    # Dictionaries
    dictionary1 = {'name': 'Drake', 'age': 15, 'home': 'sandown'}  # makes a dictonary 'dictonary1' (editable)
    # 'name' corresponds to 'Drake', 'age' corresponds to 15.....etc..
    dictionary1['planet'] = 'earth'  # adds entry 'planet' and value as 'earch'
    del dictionary1['home']  # deletes item from dictionary
    print('dictionary:', dictionary1['name'])  # prints name from dictionary1
    # for looping through all of the items go down to looping


def convertData():
    # conversion
    toConvert = [1, 1, 2, 3, 3, 'four', 'five', 'six', 'six', 'six']
    newTuple = tuple(toConvert)  # converts list to a tuple ()
    newSet = set(toTuple)  # converts to a set {}
    newList = list(toConvert)  # converts to list []

    toDict = [['one', 1], [2, 'two'], [3, 3]]  # a nested list
    newDict = dict(toDict)  # the nested items first item well be the key.
    # e.g.  {'one':1, 2:'two', 3:3}


# 2. statements, functions#		##########################################################################################################################################################

def ifElifElse():
    # 2.1 if/elif/else
    print('if/elif/else.........................')
    y = 1
    if y == 1:  # checks if y = 1
        print('1')
    elif y == 2:  # checks if y = 2
        print('2')
    else:  # if y not equal 1 or 2, does this
        print('3')
    # don't need elif, it just adds more options
    print("2.1 if/else/elif .....................................")

    # 2.2 try/except/finally
    print('try/except/finally...........................')
    inp = input("Input number: ")
    try:
        inp = int(inp)  # tries to convert to integer
        print(inp)
    except:  # if the try part fails, then this runs
        # you can have multiple except statments. to catch different errors. lookup exception errors for errors to print out
        print('not a integer!')  # if fails does this
    else:
        print('um')
    finally:  # runs at end, whether it faild or not
        print("done")
    # usually don't need finally

    try:
        pass  # code
    except ValueError:
        pass  # if exception is ValueError..
    else:
        pass  # if it's anything else then...


def functions():
    # 2.3 functions
    def adding(num1, num2):  # defines(makes) a function called adding, with paremeters 'num1','num2'
        sum = num1 + num2  # whatever you input this adds them together and puts it in 'sum'
        return sum  # returns the value of sum, becuase a function is closed off. As in outside data can't be read from inside the function and outside can't read data from within function, unless you use global or return.

    # returns 10, not print to print it >
    print(adding(5, 5))


def classes():
    # 2.4 class
    class Class1:
        def __init__(self):
            self.printHi()  # when class is called, this runs, and can call any other functions from later on in the class

        def printHi(self):
            print('Class1:', "hi")

        def printBye(self):
            print('Class1:', "bye")

    newInstance = Class1


# creates a new instance of a class, since you have a __init__ method that well run when a new instance is made

def basicListMethods():
    # 2.5 built in functions/methods
    Numbers = [2, 5, 46, 74, 3, 6, 3, 6, -3, 6, 3, 6, 2, 21, 34, 6, 8, 6, 3, 533, 34725, -12,
               -15, -15, 2]  # makes a list of numbers to play with
    # this works with tuples too
    print('min: ', min(Numbers))  # prints the minimum number in the list
    print('max: ', max(Numbers))  # prints the maximum number in the list
    print('abs: ', abs(Numbers[8]))  # prints the absolute value for the number
    print('type:', type(Numbers))  # prints what 'Numbers' is, which is list object
    print('list: ', list(Numbers))  # makes into list. but already a list
    print('tuple: ', tuple(Numbers))  # makes into tuple, similar to list main difference not editable
    print("sorted: ", sorted(
        Numbers))  # sorts the data, numbers lowest to highest(lowest negative then highest positive, -10, 1), works with letters to a-z, A-Z, etc
    print('set: ', sorted(set(
        Numbers)))  # makes set, similar to tuple but removes duplicates. there's also a sorted method on it. so it converts to set then sorts
    numDict = {k: v for k, v in zip(range(len(Numbers)), Numbers)}  # creates a dictionary, zip(keys, values)
    print('numDict: ', numDict)


# basic python methods for lists

def compareing():
    # 2.6 comparators and, not, or
    var1 = 1
    var2 = 2
    var3 = 3
    if var1 and var2 == 1:  # checks to see if 'var1' and 'var2' is 1, if one of them is not 1 then it returns false, and doen't run the code. both have to be 1
        print('both are equal to one')
    elif var2 or var3 == 2:  # if one of the var(s) is equal to 2, then it runs the code
        print("one of them is 2")
    elif not var3 == 3:  # if 'var3' not equal to 3 then is runs the code, if it does it skips
        print("var3 is not 3")
    else:
        pass  # pass does nothing, just skips


# python comparing statments

def breakContinue():
    # 2.7 break, continue
    for n in range(2, 10):  # runs through the numbers 2-10
        for x in range(2, n):  # runs through the number 2-n (8)
            if n % x == 0:  # checks to see if n % x has no remienders
                print(n, 'equals', x, '*', n // x)
                break
        else:
            # loop fell through without finding a factor
            print(n, 'is a prime number')  # if so then prints this

    print("2.7 break......")
    for num in range(2, 10):
        if num % 2 == 0:
            print("Found an even number", num)
            continue
        print("Found a number", num)


# break/continue keywords for skipping a loop or stopping

# 3. looping#		##########################################################################################################################################################
def looping():
    # 3.1 while loop
    x = 0
    while x < 5:
        print("whatup")
        x += 1
    else:
        print("done")

    # you can also add a else statement at the end if want to(doesn't need an else)
    # 3.2 function loop with for
    def whatup2():
        print("whatup2")

    print("3.2 looping. function .......................................")

    # 3.3 makes for loop to call function that prints 'whatup2'
    for i in range(5):
        whatup2()
    # calling the function 5 times
    print("3.3 looping function ...............................")

    # 3.4 for loop
    for i in range(5):
        print(i)

    # Dictionary looping
    dictToLoop = {1: 1, 'two': 2, 3: 'three', 'four': 'four'}
    for k, v in dictToLoop.items(): print('key:', k, ':: Value:', v)
    # counts down from 5 and prints it out

    x = [1, 2, 3, 4, 5]
    for i in x:
        print(i)
    else:
        print('Done')

    for i in x:
        print(i)
        if i == 3: break
    else:
        print('done')

    def find(list, target):
        for i, item in enumerate(list):  # runs through list
            if item == target: break  # if i == target it breaks the loop
        else:
            return -1  # if break was not called, it'll return -1, which means not in list
        return i  # if break is called then it'll return, which is the position in list where it stopped becuase the break

    names = ['Kirby', 'Drake', 'Emoji', 'Shades', 'Thing']
    index = find(names, 'Shades')
    print("Shades at index {} in names".format(index))


# Looping, while loop, for loop

def dictionaries():
    pass


# dictionary
def strings():
    person = ['Kirby', 10, 'Pink', 'Cute', 'Cake']

    # the S| , S3.1| repeat| is to track what corresponds with what in the code, since this prints a lot, S is sentence if you couldn't guess :-)
    sentence = "S| Hello I'm " + person[0] + ", I am " + str(person[1]) + " years old, I'm " + person[2] + ", also " + \
               person[3] + ". My favorite food is " + person[4]
    # not the bst way to print something like this out. It's hard to read, to keep track of what variable your getting, and it's somewhat long(not the sentence, it's how you wrote the format)
    print(sentence)

    sentence2 = "S2| Hello I'm {}, I am {} years old, I'm {}, also {}. My favorite food is {}.".format(person[0],
                                                                                                       person[1],
                                                                                                       person[2],
                                                                                                       person[3],
                                                                                                       person[4])
    # in the " " use {} to use as a place-holder, then .format to tell what to put in the place holder. it'll put the corresponding data according to start-end positioning
    # this is much easier to read; since you do't have the "" cut of every other part. but it's still long.
    # you can also see here with person[1] which is a integer; we didn't need to convert that into a string from .format()
    print(sentence2)

    # --- Specify position ---
    sentence2_1 = "S2.1| Hello I'm {4}, I am {0} years old, I'm {3}, also {1}. My favorite food is {2}.".format(
        person[0], person[1], person[2], person[3], person[4])
    # it looks similar to the one above, but here you can specify what data corresponds to what. In the .format() {0} would be the index of 0 in .format() duh. no you cannot change this
    # this works with dictionaries and tuples, with dictionaries you'll use the key instead, duh
    print(sentence2_1)

    # --- Repeating same data ---
    # the {1} specifying what corresponds to what is most useful if you have to repeat something, Ex.
    repeat = 'repeat| {0} is my favorite, {0} is yummy, {0} is awesome, {1} is ok but {0} is better!'.format("Cake",
                                                                                                             "Cupcake")
    # as you can see you can call again the place-holder multiple times
    print(repeat)

    # --- Grab within {} ---
    sentence3 = "S3| {0[0]} is here, {0[0]} is Ready!. {0[4]} is my favorite!".format(person)
    # here you can see that we can get data from the list from right in the {} place-holder chars, and we can repeat the {0} like from above,
    # this would be the same as > sentence3 = "{0} is here, {0} is Ready!. {1} is my favorite!".format(person[0], person[4]) . but this way we need two args in .format(), the above one we don't
    print(sentence3)

    # the {} works with class obj too, you need a class instance first, (inst = Class()) . then you can call that class object like how you would regularly >
    # sentence = "hi {}".format(inst.name) , or sentence = "hi {0.name}".format(inst)
    # you cannot just call the intire object/var/etc from with {} in "", e.g. s = "Hi {inst.name}".format() This won't work, nor well it work for variables s = "Hi {x}".format(), and no removing the .format() won't change anything

    # --- args ---
    sentence3_1 = "S3.1| {name} is here, {name} is Ready!. {food} is my favorite!".format(name="Kirby", food="Cake")
    # this well match up the names fro within {} with the args from the .format(), so {name} well be "kirby" since in .format(name="kirby")
    print(sentence3_1)

    # --- With Dict ---
    aboutPerson = {'name': 'Kirby', 'food': 'Cake', 'mood': 'happy'}  # creates dictionary about person
    sentence3_2 = "S3.2| {name} is here, {name} is Ready!. {food} is my favorite! I'm {mood} right now.".format(
        **aboutPerson)
    # the **aboutPerson in .format() is unpacking all the data from the dict, then in the {} you call the key from dict. Seems very useful
    print(sentence3_2)

    # --- {} Padding and Numbers---
    for i in range(0, 11): print("Numbers: {:02}".format(i))
    # this well only show 0 before single digit numbers. e.g. 01, 02, 03 .... 09, 10, 11
    # if you did {:03} then it'll be 001, 002, 003 .... 009, 010, 011 ... 099, 100, 110  , and so on

    print("Comma Number|{:,}".format(1000 ** 2))  # 1000**2 = 1,000,000
    # the {:,} well add a comma to seperate the section so it's more human readable instead of just 1000000 > 1,000,000  . it won't work with everything e.g. {:.} won't work >
    # what well work {:,} ,/_/+  . the + well um

    print("Decimal Number| {:.2f}".format(3.14159265359))
    # {:.2f} well only show 2 numbers after the decimal . . the f is for float
    # so {:.5f} would be 3.14159

    # ----- Datetime -----
    # Date = dt.datetime(2017, 3, 15, 11, 35, 55) # sets date and time. Year, Month, Day, Hour, Minute, Second
    # you can input your own data like above, put i'm going to use the current relavent date and time
    Date = dt.datetime.now()
    print("Datetime|", Date)  # prints out date and time in this format Y-M-D H:M:S

    print("Datetime Formatted| {:%B %d %Y, %X}".format(Date))  # formats the date as specified,
    # common ones to use. Month  -- %B - full month name (January) %b small Month name (Jan), %j - the day number of the year (50 or 360)
    # Day  --  %d - day of moth, %A full Day name(Monday), %a shortened Day name(Mon)
    # Year  --  %Y - full year (2017), %y - shortened year (17)
    # Time  --  %X - Time (11:35:55), %H - 24h Hour, %I - 12h Hour, %M - Minutes, %S, Seconds
    # All  --  %c everything (Wed Mar 15 11:35:55 2017), %x - date (03/15/17),

    print(
        "Datetime Formatted with text| Today is {0:%A} on {0:%B %d}, the {0:%j} day of the year. And the time is {0:%X}".format(
            Date))
    # you can add text and move the data around like this to give a full sentence with the date and time. you do need the 0 part in {0:%j}, since that says get data from Date

    # go to small modules.py for more on datetime


# better print output, and more complex concepts
basicMethods()

"""
~False               ~def                 ~if                  raise
~None                del                 ~import              ~return
True                ~elif                ~in                  ~try
~and                 ~else                is                  ~while
as                  ~except              lambda              with
assert              finally             nonlocal            yield
break               ~for                 ~not
~class               from                ~or
continue            ~global              ~pass

"""

File: Java/src/Learn/Basics.java
package Learn;
/**
 * Created by drake on 7/29/17.
 */

import java.util.InputMismatchException;
import java.util.Scanner;
public class Basics {
    public static void main(String args[]){

        //TODO ===== Data Types =====
        System.out.println("===== Data Types =====");
        System.out.println("Hello World");

        // Basic variables
        //String emoji = 'July 28';
        byte Byte1 = 1;             // 8bit -- -128 - 127 , 0 , useful to save memory when it's important
        short Short1 = 2;           // 16bit -- -32,768 - 32,767 , 0 ,
        int Interger1 = 28;         // 32bit -- -2^32 - 2^32-1 , 0 ,
        long Long1 = 666L;          // 64bit -- -2^63 - 2^64-1 , 0L , used for very big numbers
        float Float1 = 666.6f;      // 32bit -- 3.4e+038 , 0.0f ,
        double Double1 = 28.0d;     // 64bit -- 1.7e+308 , 0.0d , use this one the most
        char Char1 = 'h';           // 16bit -- '\u0000' -  '\uffff' (or 65,535 inclusive) , '\u0000' ,
        boolean Boolean1 = true;    // 1bit information, but size unknown , false , true/false
        String String1 = "Emoji Movie, July 28!";
                                            // xbit -- x - x , null , String can contain almost anything, ascii
        // / Print variable
        System.out.println(Interger1);

        // Declaring Variables
        int int1 = 1, int2 = 2;
        // this declares multiple int variables

        String str1, str2;
        //private String str3; String str4;
        // make private variables
        String str5; int int3;
        str1 = str2 = "Lol";
        // this creates two String vars and then sets str1 equal to str2 which is ewal to "Lol"



        // --- More Complicated Data Types ---
        // Arrays

        int[] intArray1;
        // Integer array, can only hold integers

        // Go to Basic Concepts for more, and array methods


        //TODO ===== User Input =====
        // go to Java_Utils file for more on Scanner module

        System.out.println("===== User Input =====");

        Scanner usrInp = new Scanner(System.in);
        // you need to import the Scanner module to get user input from console. then this code makes a new object to catch the data and puts into a var if so choose

        System.out.print("Input something > ");
        String usrInput1 = usrInp.nextLine();
        // calls Scanner object to get data and then prints said data on new line
        // this gets user input and stores in usrInput1 String variable to be called later
        System.out.println(usrInput1);

        System.out.print("Input Integer please > ");
        double usrInputInt1 = usrInp.nextInt();
        // depending on what input you're asking for and going to get you can use the corresponding type with nextLine, nextInt, nextDouble, etc




        // ----- if/else & Switch -----


        System.out.println("----- if/else & Switch -----");

        // --- simple if/else statement ---
        boolean ifVar = true;
        if (true == ifVar){
            System.out.println("ifVar is True");
        } else {
            System.out.println("ifVar is False");
        }



        if (ifVar) {System.out.println("ifVar is True, still"); }
        // this shows you can do things on one line, but probably shouldn't since it doesn't look very good
        // and if you want to check if a boolean var is true you don't need var == true, just put the var name in () of the statement
        // and since this if statement doesn't have an else it'll just do nothing if not true


        boolean ifVar2 = false;

        if (ifVar && ifVar2) {
            System.out.println("Both are True");
        }
        // this checks if both are true, if ifVar and ifVar2 are True then ...
        if (ifVar || ifVar2) {
            System.out.println("At least One is True");
        }
        // one or both are true works, if ifVar or ifVar2 is True then ...

        if (!ifVar2) {
            System.out.println("ifVar2 is False");
        }
        // the ! means opposite, so if ifVar2 not True then ...


        // --- Nested if/else ---
        int ifVar3 = 15;
        if (1 < ifVar3){
            System.out.println("Bigger then 1");
            if (10 < ifVar3){
                System.out.println("Bigger then 10");
            }
        }



        // --- if/else if/else ---
        int ifVar4 = 1;
        if (1 == ifVar4){
            System.out.println("One");
        }
        else if (2 == ifVar4) {
            System.out.println("Two");
        }
        else if (3 == ifVar4) {
            System.out.println("Three");
            System.out.println("...");
        }
        else {
            System.out.println("Unknown");
        }
        // if one of the if statements is true and runs then it doesn't check the others


        // --- Switch Statement ---
        System.out.println("----- Switch Statement -----");
        int switchVar1 = 1;

        switch (switchVar1){
            case 1: System.out.println("One");
            break;
            // checks if var is 1 then runs this code
            case 2:
                System.out.println("Two");
            break;
            // the break just means to stop looking at the other cases if already found matching
            case 3:
                System.out.println("Three");
                System.out.println("...");
            break;
            //  the above example just shows more compact way if you don't have a lot of code
            default: System.out.println("Unknown");
            break;
            //  this will run if it doesn't mach any of the cases
        }



        // --- Conditionals & Conditional Statements ---
        System.out.println("----- Conditional Statements -----");


        int a = 4, b = 5;
        boolean result;
        result = a < b;             // true
        result = a > b;             // false
        result = 4 >= a;            // a smaller or equal to 4 - true
        result = 6 <= b;            // b bigger or equal to 6 - false
        result = a == b;            // a equal to b - false
        result = a != b;            // a is not equal to b - true
        result = a > b || a < b;    // Logical or - true
        result = (3 < a) && (a < 6);    // Logical and - true
        result = !result;           // Logical not - false
        a += b;                     // adds, same as a = a + b
        a *= b;
        a -= b;
        a /= b;
        a %= b;
        a++;                        // adds one after a
        ++a;                        // adds before
        b--;                        // Ex. .println(a++)  output >> 4
        --b;                        // .println(++a)  output >> 5

        String a1 = "Wow";
        String b1 = "Wow";
        String sameA1 = a1;


        boolean r1 = a1 == b1;      // This is false, since a and b are not the same object
        boolean r2 = a1.equals(b1); // This is true, since a and b are logically equals
        boolean r3 = a1 == sameA1;  // This is true, since a and sameA are really the same object

        boolean conditionalVar = true;
        System.out.println(conditionalVar ? "Var is True" : "Var is False");
        // prints one thing if variable is true and something else if false

        boolean conditionalVar2 = !conditionalVar;
        // makes this variable false if first variable is true and true of false
        System.out.println(conditionalVar2);
        // you could use, boolean conditionalVar2 = !conditionalVar
        // but this is just to show conditional statements


        //TODO ===== try/catch =====
        System.out.println("===== try/catch =====");
        Scanner usrInp2 = new Scanner(System.in);

        while (true) {
            try {
                System.out.print("try/catch | Input Integer Please > ");
                int usrInput2 = usrInp2.nextInt();
                System.out.println(usrInput2);
                break;
                // tries to run this code
            } catch (InputMismatchException e) {
                System.out.println("ERROR");
                usrInp2.next();
            }
            // if try code fails then this catches and runs code
        }



        //TODO ===== Looping =====
        System.out.println("===== Looping =====");

        // --- Simple While Loop ---
        System.out.println("--- Simple While Loop ---");

        int whileInt1 = 0;
        while (10 > whileInt1){
            System.out.print(whileInt1);
            whileInt1++;
        }
        System.out.println();
        // add spacing


        // --- Do While Loop ---
        System.out.println("--- Do While Loop ---");
        // this is similar to a while loop but it loops once before checking the statement to run or not

        int doWhileVar1 = 10;
        do {
            System.out.println(doWhileVar1);
            doWhileVar1++;
        } while (10 > doWhileVar1);



        // --- For Loop ---
        System.out.println("--- For Loop ---");

        for (int i = 0; 10 > i; i++) {
            System.out.println(i);
        }
        // (makes int var; loop while i < 10; add 1 each loop) { print i }

        // go to Basic concepts for more foreach loop. for (obj i: list) { }

        // --- foreach loop ---
        System.out.println("--- forach loop ---");
        Object[] objArr1 = {"Movie", "FGSP", 28, "Emoji", 2017.7};
        for (Object i:objArr1){
            System.out.println(i);
        }

        // --- break & continue ---
        System.out.println("--- break & continue ---");

        for (int i = 0; 10 <= i; i++){
            // loops until break, watch out using while loop like this, you might get an infinite loop
            System.out.println("Beginning Loop");
            // adds one to variable, if not then infinite loop.

            if (5 <= i) {
                System.out.println("STOPPING");
                break;
            }
            // checks statement, if true it stop the loop (breaks out of it)

            System.out.println("Middle");

            if (2 == i) {
                System.out.println("Skipping Loop :(");
                continue;
            }
            // checks statement, if true it'll instantly go to the next loop, won't run any code after the continue
            // if you're doing this with a while loop, watch where you put the i++; , since if it's after the continue it won't be able to change the number and it'll always equal 2
            System.out.println("End of Loop!");
        }



        //TODO ===== Importing =====
        System.out.println("===== Importing =====");

        secondFile secondFileObject = new secondFile();
        // this creates a new secondFile object in this main file, so you call and get data from the other file. basically treat as a class
        secondFileObject.printHi();





        //TODO ===== Math =====
        System.out.println("===== Math =====");


        System.out.println("Add 1+1: " + 1+1);
        System.out.println(("Sub 1-2: ") + (1-2));
        System.out.println(("Times 2*2: ") + 2*2);
        System.out.println(("Divide 10/2: ") + 10/2);
        System.out.println("Power 10^2: " + Math.pow(10, 2));
        System.out.println("Mod 10%3: " + Math.floorMod(10, 3));
        System.out.println("Floor 10//3: " + Math.floorDiv(10, 3));
        System.out.println("Floor 10//3: " + Math.floorDiv(10, 3));
        System.out.println("Abs -10: " + Math.abs(-10));
        System.out.println("Min 1,2: " + Math.min(1, 2));
        System.out.println("Max 1,2: " + Math.max(1, 2));
        System.out.println("sqrt 25: " + Math.sqrt(25));
        System.out.println("Ceil 1.5: " + Math.ceil(1.5));
        System.out.println("Round 1.5: " + Math.round(1.5));
        System.out.println("Random: " + Math.random());
        System.out.println("Pi: " + Math.PI);
        System.out.println("e: " + Math.E);
        System.out.println("e^pi*i: " + Math.pow(Math.E, (Math.PI * Math.sqrt(-1))));


        //TODO ===== Functions =====


        System.out.println("===== Functions =====");
        Basics.func1();
        Basics.func2("July", "28");


        // for more on functions and classes go to Basic Concepts or Advacne
    }

    public static void func1() {
        System.out.println("This is func1");
    }
    // simple function, no output(return (print is not output)) and can be called without making a instance, and it's public

    public static void func2(String fname, String lname){
        System.out.println("Hello " + fname + " " + lname);
    }
    // this takes two args and prints out.
    // you have to specify data type when making parameter, int varName, String varName, etc








}
